Content.Server/Abilities/Psionics/PsionicAbilitiesSystem.cs

```diff
@@ -62,66 +77,185 @@ public void AddPsionics(EntityUid uid, bool warn = true)
+                    continue;
 
-            AddComp<PsionicComponent>(uid);
+                if (psionic.ActivePowers.Contains(powerPrototype))
+                    newPool.Weights.Remove(powerPrototype.ID);
+            }
 
-            var newComponent = (Component) _componentFactory.GetComponent(powerComp);
-            newComponent.Owner = uid;
+            if (newPool.Weights.Keys != null)
+            {
+                var newPower = _prototypeManager.Index<PsionicPowerPrototype>(newPool.Pick());
+                InitializePsionicPower(uid, newPower);
+            }
 
-            EntityManager.AddComponent(uid, newComponent);
+            _glimmerSystem.Glimmer += _random.Next(1, 5);
         }
 
-        public void AddRandomPsionicPower(EntityUid uid)
+        public void InitializePsionicPower(EntityUid uid, PsionicPowerPrototype proto, bool playPopup = true)
         {
-            AddComp<PsionicComponent>(uid);
+            if (!TryComp<PsionicComponent>(uid, out var psionic)
+                || !_prototypeManager.TryIndex<PsionicPowerPrototype>(proto.ID, out var _))
+                return;
 
-            if (!_prototypeManager.TryIndex<WeightedRandomPrototype>("RandomPsionicPowerPool", out var pool))
+            psionic.ActivePowers.Add(proto);
+
+            foreach (var id in proto.Actions)
             {
-                _sawmill.Error("Can't index the random psionic power pool!");
-                return;
+                EntityUid? actionId = null;
+                if (_actions.AddAction(uid, ref actionId, id))
+                {
+                    _actions.StartUseDelay(actionId);
+                    psionic.Actions.Add((id, actionId));
+                }
             }
 
-            // uh oh, stinky!
-            var newComponent = (Component) _componentFactory.GetComponent(pool.Pick());
-            newComponent.Owner = uid;
+            if (proto.AmplificationModifier != 0)
+                psionic.AmplificationSources.Add(proto.Name, proto.AmplificationModifier);
 
-            EntityManager.AddComponent(uid, newComponent);
+            if (proto.DampeningModifier != 0)
+                psionic.DampeningSources.Add(proto.Name, proto.DampeningModifier);
 
-            _glimmerSystem.Glimmer += _random.Next(1, 5);
+            if (playPopup)
+                _popups.PopupEntity("generic-power-initialization-feedback", uid, uid, PopupType.MediumCaution);
+            // TODO: Replace this with chat message: _popups.PopupEntity(proto.InitializationFeedback, uid, uid, PopupType.MediumCaution);
+
+            if (proto.Components is not null)
+                foreach (var comp in proto.Components)
+                    if (!EntityManager.TryGetComponent(uid, _componentFactory.GetComponent(comp).GetType(), out var powerComp)
+                        && powerComp is not null)
+                        AddComp(uid, powerComp);
```

Comment:
A few braces could make this block of code a lot more readable

Reply:
Our current code style asks that we only use braces for multiple lines under indents, but I think I can probably make a compromise here by following your suggestion to split the individual foreach loops into functions, and then documenting said functions.